src/librustc_middle/ty/error.rs

```diff
@@ -58,6 +58,8 @@ pub enum TypeError<'tcx> {
     ConstMismatch(ExpectedFound<&'tcx ty::Const<'tcx>>),
 
     IntrinsicCast,
+    /// Safe `#[target_feature]` functions are not assignable to safe function pointers.
+    TargetFeatureCast(Box<[Span]>),
```

Comment:
I suspect it would be enough to store the DefId of the function item, the error reporting code can get those spans from the DefId. Does that make sense to you or did I miss anything?

To be clear: I'd consider this an optional clean-up, I don't want to hold up the PR over it, but it would be good to leave a TODO comment so the idea doesn't get lost entirely.